File: Sources/GatewayCore/SessionPlacement.swift
import Foundation

/// Repairs sessions written under their starting level by builds that confused
/// a template's `@level` with the destination reached by its level cues.
///
/// The operation only moves whole render directories and updates the manifest.
/// It never rewrites audio, drops notes, or replaces an existing destination.
public enum SessionPlacement {
    public struct Repair: Equatable, Sendable {
        public var track: String
        public var from: String
        public var to: String
    }

    public enum Failure: Error, LocalizedError, Equatable {
        case destinationExists(String)

        public var errorDescription: String? {
            switch self {
            case .destinationExists(let path):
                "session placement repair will not replace existing data at \(path)"
            }
        }
    }

    @discardableResult
    public static func repair(library: Library,
                              fileManager fm: FileManager = .default) throws -> [Repair] {
        var repairs: [Repair] = []
        for folder in library.focus {
            for track in folder.renders {
                let manifestURL = track.appending(path: "manifest.json")
                guard var manifest = SessionManifestIO.load(manifestURL) else { continue }
                // **A continuous journey is already where it belongs.**
                //
                // This pass asks `sessionDestination`, which answers with a
                // *documented* level -- the right answer for an authored tape,
                // which is always filed under one. A journey may arrive at a
                // station nothing describes, and `sessionDestination` can
                // never name it: asked about a journey to F13 it says F12,
                // the last documented level on the way.
                //
                // So this pass took a journey correctly written to `focus/F13`
                // and moved it to `focus/F12`, rewriting its manifest to
                // agree. The player then looked under F13, found nothing, and
                // reported "no session.wav yet" -- a repair breaking the one
                // kind of session whose destination was never in doubt,
                // because the journey carries its target explicitly.
                guard manifest.purpose != .continuousJourney else { continue }
                guard let destination = library.sessionDestination(
                        startLevel: manifest.startLevel, cues: manifest.cues)?.key else { continue }

                let current = folder.key
                if current == destination {
                    if manifest.level != destination {
                        manifest.level = destination
                        try SessionManifestIO.save(manifest, to: manifestURL)
                    }
                    continue
                }

                let parent = library.root.appending(path: "focus/\(destination)/renders")
                let target = parent.appending(path: track.lastPathComponent)
                guard !fm.fileExists(atPath: target.path) else {
                    throw Failure.destinationExists(target.path)
                }
                try fm.createDirectory(at: parent, withIntermediateDirectories: true)
                try fm.moveItem(at: track, to: target)
                manifest.level = destination
                try SessionManifestIO.save(manifest,
                    to: target.appending(path: "manifest.json"))
                repairs.append(Repair(track: track.lastPathComponent,
                                      from: current, to: destination))
            }
        }
        return repairs
    }
}

File: cross-platform/src/core/sessionPlacement.ts
/**
 * Repairs sessions written under their starting level by builds that confused
 * a template's `@level` with the destination reached by its level cues.
 *
 * The operation only moves whole render directories and updates the manifest.
 * It never rewrites audio, drops notes, or replaces an existing destination.
 */
import { existsSync, mkdirSync, readFileSync, renameSync, writeFileSync } from "fs";
import { basename, join } from "path";
import { sessionDestinationFromCues, type Library } from "./library.js";
import { encodeManifest, loadManifest } from "./sessionManifest.js";

export interface Repair { track: string; from: string; to: string }

export class PlacementError extends Error {
  readonly kind = "destinationExists";
  constructor(readonly path: string) {
    super(`session placement repair will not replace existing data at ${path}`);
  }
}

/**
 * **A continuous journey is already where it belongs.** This asks
 * `sessionDestinationFromCues`, which answers with a *documented* level — the
 * right answer for an authored tape, which is always filed under one. A
 * journey may arrive at a station nothing describes, and that function can
 * never name it: asked about a journey to F13 it says F12, the last
 * documented level on the way. Moving a journey correctly filed under F13 to
 * F12 would break the one kind of session whose destination was never in
 * doubt, because the journey carries its target explicitly.
 */
export function repair(library: Library): Repair[] {
  const repairs: Repair[] = [];
  for (const folder of library.focus) {
    for (const track of folder.renders) {
      const manifestPath = join(track, "manifest.json");
      let raw: string;
      try { raw = readFileSync(manifestPath, "utf8"); } catch { continue; }
      const manifest = loadManifest(raw);
      if (manifest === undefined) continue;
      if (manifest.purpose === "continuousJourney") continue;
      const destination = sessionDestinationFromCues(
        library, manifest.startLevel, manifest.cues)?.key;
      if (destination === undefined) continue;

      const current = folder.key;
      if (current === destination) {
        if (manifest.level !== destination) {
          writeFileSync(manifestPath, encodeManifest({ ...manifest, level: destination }), "utf8");
        }
        continue;
      }

      const parent = join(library.root, "focus", destination, "renders");
      const target = join(parent, basename(track));
      if (existsSync(target)) throw new PlacementError(target);
      mkdirSync(parent, { recursive: true });
      renameSync(track, target);
      writeFileSync(join(target, "manifest.json"), encodeManifest({ ...manifest, level: destination }), "utf8");
      repairs.push({ track: basename(track), from: current, to: destination });
    }
  }
  return repairs;
}
